### PR TITLE
feat(mpz-common): Context::blocking

### PR DESCRIPTION
**Pull request by sinui0**
This PR adds the `Context::blocking` method which supports temporarily moving a context to a separate thread to perform blocking computation. I've also moved the CPU backend shim abstraction into `mpz-common`.

This method is needed to avoid deadlocks on single threaded machines. The alternative approach is to use an `mpsc` channel between the worker thread and the calling context, but that doesn't work if single-threaded.

### CODE REVIEW

**Review by th4s — COMMENTED** [review]
Hmm, I am not sure I understand this PR. A few questions:

> This method is needed to avoid deadlocks on single threaded machines.

What is a single-threaded machine and why do we need to support single-threaded machines? Or do you mean a machine with a single CPU core?

> The alternative approach is to use an mpsc channel between the worker thread and the calling context, but that doesn't work if single-threaded.

Why doesn't this work?

**Comment by sinui0** [comment]
> What is a single-threaded machine and why do we need to support single-threaded machines? Or do you mean a machine with a single CPU core?

Yes, I'm referring to single-core CPUs, and generally environments without threading. Not to be confused with our logical thread abstraction in `mpz-common`.

> Why doesn't this work?

It does work for `CpuBackend::blocking_async` but not for `CpuBackend::blocking`. Hard to point out exactly why, but it leads to a deadlock.

---

The primary purpose of the `Context::blocking` method is that it handles moving a context to another thread in multi-threaded case. This isn't otherwise possible because most callers are going to be working with a `&mut impl Context` which obviously isn't `'static`.

File: crates/mpz-common/src/cpu.rs [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+#[cfg(any(feature = "force-st", not(feature = "rayon")))]
+mod st {
+    use futures::Future;
+
+    /// A single-threaded CPU backend.
+    #[derive(Debug)]
+    pub struct SingleThreadedBackend;
+
+    impl SingleThreadedBackend {
+        /// Execute a future on the CPU backend.
```
Comment (themighty1) [inline crates/mpz-common/src/cpu.rs]:
```suggestion
        /// Executes a future on the CPU backend.
```

File: crates/mpz-common/src/cpu.rs [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+        #[inline]
+        pub fn blocking_async<F>(fut: F) -> impl Future<Output = F::Output> + Send
+        where
+            F: Future + Send + 'static,
+            F::Output: Send,
+        {
+            fut
+        }
+
+        /// Execute a closure on the CPU backend.
```
Comment (themighty1) [inline crates/mpz-common/src/cpu.rs]:
```suggestion
        /// Executes a closure on the CPU backend.
```

File: crates/mpz-common/src/cpu.rs [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+mod rayon_backend {
+    use futures::{channel::oneshot, Future};
+    use pollster::block_on;
+
+    /// A Rayon CPU backend.
+    #[derive(Debug)]
+    pub struct RayonBackend;
+
+    impl RayonBackend {
+        /// Execute a future on the CPU backend.
```
Comment (themighty1) [inline crates/mpz-common/src/cpu.rs]:
```suggestion
        /// Executes a future on the CPU backend.
```

File: crates/mpz-common/src/cpu.rs [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+                let (sender, receiver) = oneshot::channel();
+                rayon::spawn(move || {
+                    let output = block_on(fut);
+                    _ = sender.send(output);
+                });
+                receiver.await.expect("worker thread does not drop channel")
+            }
+        }
+
+        /// Execute a closure on the CPU backend.
```
Comment (themighty1) [inline crates/mpz-common/src/cpu.rs]:
```suggestion
        /// Executes a closure on the CPU backend.
```

**Review by themighty1 — APPROVED** [review]

**Review by th4s — APPROVED** [review]

**Pull request merged**